libgnucash/core-utils/gnc-filepath-utils.cpp

```diff
@@ -1017,6 +1018,24 @@ gnc_userdata_dir_as_path (void)
     return gnc_userdata_home;
 }
 
+gchar *gnc_file_path_absolute (const gchar *prefix, const gchar *relative)
+{
+    bfs::path path_relative (relative);
+    path_relative.imbue (bfs_locale);
+    bfs::path absolute_path;
+
+    if (prefix == nullptr)
+        absolute_path = absolute (path_relative, gnc_userdata_dir_as_path());
```

Comment:
While I agree with you gnc_user_data_dir is less random than working directory, it's not the current file's directory. In addition the GNC_USERDATA_DIR is a hidden directory (or at least well hidden) so it's not really a very user friendly solution.

However thinking about this I do understand my request comes with a few challenges. Allow me to think out loud a bit here:

* The mysql and postgres backends don't have a file's directory. So what to do for those ?
* We could state that for those backends relative paths and an empty prefix are not allowed.
* But then what to do if one converts a file to a db ?
* In that case we could run a conversion if the prefix is empty.
* File->Db: replace the empty prefix with the directory of the file being saved from.
* If we can guarantee a non-empty prefix like that while converting to db, the inverse wouldn't be a problem. If we convert from db to file, the db should have either a prefix or no relative file associations.

The corner case(s) here are still in db land:
* if the user creates a new db, it won't have a prefix set. Which is fine as adding file associations will then store absolute paths (though we may want to add a sanity check in the lower level path setting routines to catch invalid values being passed via bindings)
* suppose the user had set a prefix in the db and has added a number of relative paths. And then suppose the user decides to remove the prefix. That would lead to invalid file paths.

I believe we have/had a similar issue to solve for the transaction logs in the sql db's. The current solution is to store those in GNC_USERDATA_DIR/translog. Perhaps it's time we bring some order and logic into this all for gnucash 4.x (other work in that direction is ongoing with saved-reports).

Until we can do so, perhaps it's better to not allow relative file associations with an empty prefix. And if you can do that, the default path you have chosen here is less important.

Reply:
Using a database was the reason why I chose the user data directory, I did not want to make it more complicated and it really is only used in the situation when the path_head is not set.

This can happen currently if it was set and then cleared, (Note: this would be fixed by your other suggestion of scanning for existing relative paths and making them absolute, possibly with a question to the user) or buy using the file on another computer with the preference not set and so to make valid paths the default is used. These paths are obviously incorrect but allow the dialogs to display this and hopefully the user will notice and correct.

I think that absolute paths should be stored unless the user has set the path head, at least they have made a conscious decision to do so. I suppose the interface could be improved by highlighting the path head is not set and so relative paths will not be used, I can look at that when I create the PR for invoice associations.

 Currently if it is not set, then you can not store relative paths so absolute paths will be stored.